L1Trigger/TrackFindingTMTT/src/ChiSquared4ParamsApprox.cc

```diff
@@ -0,0 +1,221 @@
+        Vinv(2 * i + 1, 2 * i + 1) = 1 / stubs_[i]->sigmaZ();
+      } else {
+        Vinv(2 * i, 2 * i) = 1 / stubs_[i]->sigmaZ();
+        Vinv(2 * i + 1, 2 * i + 1) = 1 / stubs_[i]->sigmaX();
+      }
+    }
+    return Vinv;
+  }
+
+  std::vector<double> ChiSquared4ParamsApprox::residuals(std::vector<double> x) {
+    unsigned int n = stubs_.size();
+
+    std::vector<double> delta;
+    delta.resize(2 * n);
+
+    std::map<std::string, double> trackParams = vecToMap(x);  // Get the track params by label
+    double rInv = trackParams["rInv"];
+    double phi0 = trackParams["phi0"];
+    double t = trackParams["t"];
+    double z0 = trackParams["z0"];
+
+    double chiSq = 0.0;
+
+    unsigned int j = 0;
+
+    if (getSettings()->debug() == 6)
+      std::cout << "Residuals (" << chiSq << ") [" << getSettings()->invPtToInvR() / rInv << "]: ";
+
+    largestresid_ = -1.0;
+    ilargestresid_ = -1;
+
+    for (unsigned int i = 0; i < n; i++) {
+      double ri = stubs_[i]->r();
+      double zi = stubs_[i]->z();
+      double phii = stubs_[i]->phi();
+      const double sigmax = stubs_[i]->sigmaX();
+      const double sigmaz = stubs_[i]->sigmaZ();
+
+      if (stubs_[i]->barrel()) {
+        //we are dealing with a barrel stub
+
+        double halfRinvRi = 0.5 * ri * rInv;
+        double aSinHalfRinvRi = halfRinvRi + 0.67 * halfRinvRi * halfRinvRi * halfRinvRi;
+
+        double deltaphi = phi0 - aSinHalfRinvRi - phii;
+        if (deltaphi > M_PI)
+          deltaphi -= 2 * M_PI;
+        if (deltaphi < -M_PI)
+          deltaphi += 2 * M_PI;
+        delta[j++] = (ri * deltaphi) / sigmax;  // TODO this is different from tracklet
+        delta[j++] = (z0 + (2.0 / rInv) * t * aSinHalfRinvRi - zi) / sigmaz;
+      } else {
+        //we are dealing with a disk hit
+
+        double tInv = 1 / t;
+
+        double r_track = (zi - z0) * tInv;
+        double phi_track = phi0 - 0.5 * rInv * (zi - z0) * tInv;
+        int iphi = stubs_[i]->iphi();
+
+        // N.B. These represent HALF the width and number of strips of sensor.
+        double width = stubs_[i]->width() / 2.0;
+        double nstrip = stubs_[i]->nstrip() / 2.0;
+
+        double Deltai = width * (iphi - nstrip) / nstrip;  //A bit of a hack...
+
+        if (stubs_[i]->z() > 0.0)
+          Deltai = -Deltai;
+
+        double DeltaiOverRi = Deltai / ri;
+        double theta0 = (DeltaiOverRi) +
+                        0.67 * (DeltaiOverRi) * (DeltaiOverRi) *
+                            (DeltaiOverRi);  //+0.125*DeltaiOverRi*DeltaiOverRi*DeltaiOverRi*DeltaiOverRi*DeltaiOverRi;
+
+        double Delta = Deltai - r_track * (theta0 - (phi_track - phii));
+
+        delta[j++] = (r_track - ri) / sigmaz;
+        delta[j++] = Delta / sigmax;
+      }
+
+      if (getSettings()->debug() == 6)
+        std::cout << delta[j - 2] << " " << delta[j - 1] << " ";
+
+      chiSq += delta[j - 2] * delta[j - 2] + delta[j - 1] * delta[j - 1];
+
+      if (fabs(delta[j - 2]) > largestresid_) {
```

Comment:
As most our of .cc files have "using namespace std;", if we use std::abs(), people may be tempted to shorten it to abs(), and it is then people reading the code can be confused as to whether this is the C or std version (which give different results). I therefore have preference for fabs(), if you permit it?

Reply:
This demonstrates why "using namespace std" may not be desirable, because it leads to ambiguities. Justifying one suboptimal choice on the basis of another suboptimal choice is not a good practice.